### PR TITLE
add serialize/deserialize v3 methods

### PR DESCRIPTION
**Pull request by MarkZuber**

### CODE REVIEW

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs [hunk 1]
```diff
@@ -207,6 +207,33 @@ public void Deserialize(byte[] adalState)
             }
         }
 
+        /// <summary>
+        ///
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs]: do you need information here? #Resolved

**Review by jennyf19 — APPROVED** [review]

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs (hunk 1 above)
Comment (MarkZuber) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs]: will add documentation when we finalize the API signatures.  this is in there for Henrik to get access to the serialization methods so we can finalize our compat/migration stories #Resolved

**Comment by jmprieur** [comment]
LGTM.
I'll provide the XML comments. #Resolved

**Review by jennyf19 — APPROVED** [review]

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs [hunk 2]
```diff
@@ -207,6 +207,33 @@ public void Deserialize(byte[] adalState)
...
+                var jsonSerializer = new TokenCacheJsonSerializer(tokenCacheAccessor);
+                return jsonSerializer.Serialize();
+            }
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="bytes"></param>
+        public void DeserializeV3(byte[] bytes)
```
Comment (henrik-me) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs]: this needs an additional param, called merge (default should potentially be false as it is currently in MSAL v3) #WontFix

Comment (henrik-me) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs]: >DeserializeV3 [](start = 20, length = 13)

name, DeserializeMsalV3 #Resolved

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs [hunk 3]
```diff
@@ -207,6 +207,33 @@ public void Deserialize(byte[] adalState)
             }
         }
 
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <returns></returns>
+        public byte[] SerializeV3()
```
Comment (henrik-me) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs]: >SerializeV3 [](start = 22, length = 11)

name, SerializeMsalV3 #Resolved

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/Cache/CacheFallbackOperations.cs [hunk 4]
```diff
@@ -316,7 +316,7 @@ internal class CacheFallbackOperations
                             foreach (MsalAccountCacheItem accountCacheItem in accounts)
                             {
                                 if (rtCacheItem.HomeAccountId.Equals(accountCacheItem.HomeAccountId, StringComparison.OrdinalIgnoreCase)
-                                    && accountCacheItem.PreferredUsername.Equals(upn, StringComparison.OrdinalIgnoreCase))
+                                    && (string.IsNullOrEmpty(upn) || accountCacheItem.PreferredUsername.Equals(upn, StringComparison.OrdinalIgnoreCase)))
```
Comment (henrik-me) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/Cache/CacheFallbackOperations.cs:319]: >(string.IsNullOrEmpty(upn)  [](start = 39, length = 27)

is this the fix we allowing me to share tokens between MSAL v3 and ADAL v5? #Closed

**Comment by henrik-me** [comment]
>         public CacheData SerializeAdalAndUnifiedCache()

pointing at one thing here, however I think we should generally have the same cache API as in MSAL v3. e.g. We would want this one to go away #Resolved

---
Refers to: src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs:177 in c10fafc. [](commit_id = c10fafc475c19e66bb462a4127646af092118e5a, deletion_comment = False)

**Comment by henrik-me** [comment]
Will you be adding the remaining serialize/deserialize methods to the API as added for MSAL v3? #Resolved

**Review by henrik-me — DISMISSED** [review]
:clock1:

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/Cache/CacheFallbackOperations.cs (hunk 4 above)
Comment (MarkZuber) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/Cache/CacheFallbackOperations.cs:319]: this fixes the case where the user calls AcquireTokenSilentAsync() with no account/user information.  This check is already done in the adal cache filtering but is ignored in the msal fallback case so any msal deserialization won't find the token here if the user/account information is not specified.

---
In reply to: [260969491](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1500#discussion_r260969491) [](ancestors = 260969491)

**Comment by MarkZuber** [comment]
Do we want/need them?  I wasn't here when the other methods were discussed so let's sync up on what we want in ADAL and I can get it here as needed.

---
In reply to: [468053822](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1500#issuecomment-468053822) [](ancestors = 468053822)

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs (hunk 1 above)
Comment (MarkZuber) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs]: JM was going to do API documentation.

---
In reply to: [255733681](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1500#discussion_r255733681) [](ancestors = 255733681)

**Comment by MarkZuber** [comment]
>         public CacheData SerializeAdalAndUnifiedCache()

go away with [Obsolete] attribute or go away "remove this entirely"?  this will be adalv5 so a breaking change is fine, just clarifying what we want.

---
In reply to: [468053692](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1500#issuecomment-468053692) [](ancestors = 468053692)

---
Refers to: src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs:177 in c10fafc. [](commit_id = c10fafc475c19e66bb462a4127646af092118e5a, deletion_comment = False)

**Comment by henrik-me** [comment]
Looks great to me. 

Before commit: Thinking these changes should go into a new v5 branch? or we should branch current dev to dev4 ?

**Comment by henrik-me** [comment]
>         public CacheData SerializeAdalAndUnifiedCache()

looks great. imo, those should all be set to fail, thoughts JM?

---
In reply to: [468061232](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1500#issuecomment-468061232) [](ancestors = 468061232,468053692)

---
Refers to: src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs:177 in c10fafc. [](commit_id = c10fafc475c19e66bb462a4127646af092118e5a, deletion_comment = False)

**Comment by jmprieur** [comment]
> public CacheData SerializeAdalAndUnifiedCache()
> 
> 
> looks great. imo, those should all be set to fail, thoughts JM?
> 
> In reply to: 468061232 [](ancestors = 468061232,468053692)
> 
> Refers to: src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs:177 in c10fafc. [](commit_id = c10fafc, deletion_comment = False)

I don't understand the question, @henrik-me

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs (hunk 2 above)
Comment (MarkZuber) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs]: as discussed, we're not doing this.

---
In reply to: [260968891](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1500#discussion_r260968891) [](ancestors = 260968891)

**Comment by MarkZuber** [comment]
I would vote to create a dev4x off of dev (or create it if/when we want/need it) since the v5x changes are pretty minor compared to what we did for dev3x in msal.

---
In reply to: [468484578](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1500#issuecomment-468484578) [](ancestors = 468484578)

**Comment by henrik-me** [comment]
dev4x makes sense to me. we should branch off and ensure we have the build+test automation running on that branch ensuring we can ship a hotfix if needed.

**Comment by MarkZuber** [comment]
UI automation passed here:  https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=300867

**Pull request merged**